File: 240703/최대 감소 부분 수열/longest-decreasing-subsequence.java
import java.util.*;
import java.math.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    static int n = sc.nextInt();
    static int[] seq = new int[n+1];
    static int[] dp = new int[n+1];

    public static void main(String[] args) {
        
        for(int i = 1; i <= n; i++) {
            seq[i] = sc.nextInt();
        }

        dp[1] = 1;

        for(int i = 2; i <= n; i++){
            int maxDp = 0;
            for(int j = 1; j < i; j++) {
                if(seq[i] < seq[j]) {
                    maxDp = Math.max(maxDp, dp[j] + 1);
                }
            }
            dp[i] = maxDp;
        }

        int result = 1;
        for(int i = 1; i <= n; i++) {
            result = Math.max(result, dp[i]);
        }

        
        System.out.println(result);
    }
}

File: 240703/최대 감소 부분 수열/longest-decreasing-subsequence.kt
import java.util.*
import kotlin.math.*

fun main() {
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    val seq = IntArray(n)
    val dp = IntArray(n) { 1 } // 모든 dp 요소를 1로 초기화

    for (i in 0 until n) {
        seq[i] = sc.nextInt()
    }

    for (i in 1 until n) {
        for (j in 0 until i) {
            if (seq[i] < seq[j]) {
                dp[i] = max(dp[i], dp[j] + 1)
            }
        }
    }

    val result = dp.maxOrNull() ?: 1 // dp 배열의 최댓값이 결과

    println(result)
}